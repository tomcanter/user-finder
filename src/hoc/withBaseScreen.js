// @flow

import React from 'react';
import qs from 'query-string';
import DocumentTitle from 'react-document-title';
import isEmpty from 'lodash/fp/isEmpty';
import curry from 'lodash/fp/curry';
import {
  css,
  StyleSheet,
} from 'aphrodite/no-important';
import HeaderContainer from 'containers/Header';
import {PER_PAGE} from 'store/api';

type Props = {
  push: Function,
  location: Object,
};

export const pushUrlQuery = curry((push: Function, query: string) => {
  push({
    pathname: '/search',
    search: `per_page=${PER_PAGE}&page=1&q=${query}`,
  });
});

export function constructTitle(search: Object) {
  const base = 'Github User Search';
  if (isEmpty(search)) {return base;}
  const {q, page} = search;
  return `${q} - Page ${page} - ${base}`;
}

function getDisplayName(WrappedComponent) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}

function withBaseScreen(WrappedComponent: ReactClass<*>) {
  function component({location, push, ...restProps}: Props) {
    const parsedSearch = qs.parse(location.search);
    const title = constructTitle(parsedSearch);

    return (
      <DocumentTitle title={title}>
        <div className={css(styles.DefaultLayout)}>
          <HeaderContainer
            searchTerm={parsedSearch.q}
            searchQuery={qs.stringify(parsedSearch)}
            onSubmit={pushUrlQuery(push)}
          />
          <WrappedComponent
            searchTerm={parsedSearch.q}
            location={location}
            {...restProps}
          />
        </div>
      </DocumentTitle>
    );
  }
  component.displayName = `withBaseLayout(${getDisplayName(WrappedComponent)})`;
  return component;
}

const styles = StyleSheet.create({
  DefaultLayout: {
    flexDirection: 'column',
    display: 'flex',
    minHeight: '100vh',
  },
});

export default withBaseScreen;
