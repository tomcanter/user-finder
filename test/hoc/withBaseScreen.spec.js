import React from 'react';
import renderer from 'react-test-renderer';
import {
  StyleSheetTestUtils,
} from 'aphrodite';
import {Provider} from 'react-redux';
import {
  BrowserRouter as Router,
} from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import withBaseScreen, {
  constructTitle,
  pushUrlQuery,
} from 'hoc/withBaseScreen';

function createComponent(component) {
  const mockStore = configureMockStore();
  const store = mockStore();
  return renderer.create(
    <Provider store={store}>
      <Router>
        {component}
      </Router>
    </Provider>
  ).toJSON();
}

describe('HOC: withBaseScreen', () => {

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  describe('when passed a component', () => {
    it('should render it in the base screen', () => {
      const props = {
        location: {},
        push: jest.fn(),
        other: 'test',
      };
      const component = () => <p>test</p>;
      const WithBase = withBaseScreen(component);
      expect(createComponent(<WithBase {...props} />)).toMatchSnapshot();
    });
  });

  describe('constructTitle function', () => {
    it('should use the base title when no search query is present', () => {
      const title = constructTitle({});
      expect(title).toEqual('Github User Search');
    });

    it('should return a title describing current search and page', () => {
      const title = constructTitle({q: 'foo', page: 2});
      expect(title).toEqual('foo - Page 2 - Github User Search');
    });
  });

  describe('pushUrlQuery function', () => {
    it('should push a search query on the URL', () => {
      const spy = jest.fn();
      pushUrlQuery(spy, 'testing');
      expect(spy.mock.calls[0]).toMatchSnapshot();
    });
  });

});

