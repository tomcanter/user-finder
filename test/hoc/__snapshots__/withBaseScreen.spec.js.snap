// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC: withBaseScreen pushUrlQuery function should push a search query on the URL 1`] = `
Array [
  Object {
    "pathname": "/search",
    "search": "per_page=42&page=1&q=testing",
  },
]
`;

exports[`HOC: withBaseScreen when passed a component should render it in the base screen 1`] = `
<div
  className="DefaultLayout_1ra77ly"
>
  <div
    className="Container_10ki57k-o_O-HeaderContainer_12tblyh"
  >
    <div
      className="Container_inner_ydook-o_O-Container_padding_1e7x1pk"
    >
      <div
        className="Header_16l6lg7"
      >
        <div
          className="Header_logo_lmjs6u"
        >
          <div>
            <h1
              className="Logo_title_1cswrtu"
            >
              <a
                className="Logo_link_1pkhm2c"
                href="/"
                onClick={[Function]}
              >
                GitHub User Search
              </a>
            </h1>
            <p
              className="Logo_subTitle_h76sf7"
            >
              Browse users and their profiles via 
              <a
                href="https://developer.github.com/v3/"
              >
                the GitHub API
              </a>
            </p>
          </div>
        </div>
        <div
          className="Header_form_q7zmln"
        >
          <form
            className="SearchForm_13hv8io"
            onSubmit={[Function]}
          >
            <label
              className="hiddenVisually_1qrrm43"
              htmlFor="searchInput"
            >
              Search for a user
            </label>
            <input
              className="SearchForm_input_9ohod8"
              id="searchInput"
              onChange={[Function]}
              placeholder="Search for a user, e.g. tomcanter"
              type="text"
              value=""
            />
            <button
              className="SearchForm_btn_tixetf"
              type="submit"
            >
              <svg />
              <span
                className="hiddenVisually_1qrrm43"
              >
                Submit
              </span>
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
  <p>
    test
  </p>
</div>
`;
